feat: add releases widget

Aandree5 opened the pull request:
<br/>
<div align="center">
  <img src="https://homarr.dev/img/logo.png" height="80" alt="" />
  <h3>Homarr</h3>
</div>

**Thank you for your contribution. Please ensure that your pull request meets the following pull request:**

- [x] Builds without warnings or errors (``pnpm build``, autofix with ``pnpm format:fix``)
- [x] Pull request targets ``dev`` branch
- [x] Commits follow the [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] No shorthand variable names are used (eg. ``x``, ``y``, ``i`` or any abbrevation)

Requires: https://github.com/homarr-labs/homarr/pull/2496

The goal for this is to have a widget where we can load a series of repositories that we want to follow and show their new releases. Sometimes getting a notification is not the most useful thing because we may not be able to act on it right there and then, so with this we can at any point open the dashboard and see what new releases are available and which ones we want to act on.


https://github.com/user-attachments/assets/2c264c57-ef9a-4e04-a93a-0c2a0be0e013

    
![image](https://github.com/user-attachments/assets/1825ec7a-1642-4b61-b821-7f02ef5f9e2e)
**Left:**
    - Show Only Highlighted: **off**
    - Show Details: **on**
    
**Middle:**
    - Show Only Highlighted: **on**
    - Show Details: **off**
    
**Right:**
    - Show Only Highlighted: **on**
    - Show Details: **on**
    - 
![image](https://github.com/user-attachments/assets/e3c3bf0b-59a5-4892-8cac-583ee82b039c)

**Options page:**
![image](https://github.com/user-attachments/assets/75f97d42-210f-476a-96db-a036f95bbae5)
![image](https://github.com/user-attachments/assets/7f185e39-0f79-46d4-b6b1-b571b3dce2b6)
The `Version Regex` box is an autocomplete field, so we can have standard regex codes but still allow the user to make their own, if they want to track a specific release type.

### Edit 09 Apr 2025
**Version Filtering Changed**
`versionRegex` field has been changed to `versionFilter`, to help reduce the risk of malicious RegEx patterns

https://github.com/user-attachments/assets/a1a36dcf-0ae7-4af3-aa16-72985fe7b49e

Meierschlumpf commented:
Hmm an intresting idea, I'll need to get the opinion of the other maintainers on this. Thank you for considering a contribution ❤️

Aandree5 commented:
> Hmm an intresting idea, I'll need to get the opinion of the other maintainers on this. Thank you for considering a contribution ❤️

I understand, let me know what everyone thinks.
I do honestly like the way this tool is being developed, I'm happy to help by contributing with what I can. I was actaully about to start building a simillar solution when a friend mentioned homarr and I found it to be basically what I was enivioning.

manuel-rw commented:
Hello @Aandree5 , thanks for waiting for our reply.
We discussed this in our core team and decided that we'd love to accept your contribution.
We have not yet checked the source code, since the PR is still marked as a draft.
You can continue to work on this and mark it as "ready to review", as soon as you're done.
Please make sure that all checkboxes have been checked before you request review.

Since Homarr's purpose is not to update images but to organize and build a dashboard, we would also like the widget to not update images or recreate containers.
Let us know if you have any questions and thanks again for the contribution.

Aandree5 commented:
@manuel-rw Sounds goods to me, I'm glad you are open to contributions. I was not thinking of updating the images, I'm with you on "this is a dashboard" and is for information "only".
The question I had was if there was a an "easy" ways to remember the last version checked, this was I could highligh any version changes, if not I will revert to highlight any releases in the last `x days` configuratble in the widget settings.

Yes I will make sure the checkboxes are properly checked

Meierschlumpf commented:
I think you could use locale storage or something like that to `acknowledge` the release as seen. The only downsite would be that this will not work over multiple devices, so if you acknowledge it on your phone, it will still show on your pc

Aandree5 commented:
Sorry about the commit messages, usually there where a couple that weren't proper, I've just fixed them.
I think I also fixed the other issue, they were related with adding `react-markdown`.

Meierschlumpf commented:
No problem, in general it is not really a problem if you have some messages not following the conventional commits as they are squashed anyway to one single commit 😉

Meierschlumpf commented:
Should I review this tomorrow? Looks great so far!

Aandree5 commented:
> Should I review this tomorrow? Looks great so far!

Yes sure, o think it's all sorted, let me know if you find any issues

Meierschlumpf commented on packages/widgets/src/releases/index.ts:
```diff
@@ -0,0 +1,67 @@
+                  }
+                },
+                {
+                  message: "Invalid regular expression",
```
You can use a custom validation message the following way:

```ts
{
  params: createCustomErrorParams({
    key: "invalidRegex"
  })
}
```

instead of `message`

The translations are in `common.zod.errors.custom.*`

Aandree5 replied:
Can't get this to work, added 
```
               {
                  params: createCustomErrorParams({
                    key: "invalidRegex",
                    params: {},
                  })
                }
```
in place of message, also looked at other uses but keep getting the same error `Unsafe call of a(n) `error` type typed value.`

Aandree5 replied:
You are right, It seems to work fine now, maybe just needed a good old reboot. I will not make this change though because I'm going to work on changing to the prefix a suffix method, which will invalidate this regex field

Meierschlumpf commented on packages/widgets/src/releases/release-providers.ts:
```diff
@@ -0,0 +1,26 @@
+export class ReleaseProvider {
+  name: string;
+  iconUrl: string;
+
+  constructor(name: string, iconUrl: string) {
+    this.name = name;
+    this.iconUrl = iconUrl;
+  }
+}
```
I would suggest to just use objects

Aandree5 replied:
Ok, changed with ae4ae8542a5abf2c6a3e2e848208bfb26c89af80

Meierschlumpf commented on packages/widgets/src/releases/release-repository.ts:
```diff
@@ -0,0 +1,40 @@
+import type { ReleaseProvider } from "./release-providers";
+
+export class ReleaseRepository {
+  provider: ReleaseProvider;
+  identifier: string;
+  versionRegex: string | undefined;
+  iconUrl: string | undefined;
+
+  latestRelease: string | undefined;
+  latestReleaseAt: Date | undefined;
+  isNewRelease: boolean;
+  isStaleRelease: boolean;
+
+  releaseUrl: string | undefined;
+  releaseDescription: string | undefined;
+  isPreRelease: boolean | undefined;
+
+  projectUrl: string | undefined;
+  projectDescription: string | undefined;
+  isFork: boolean | undefined;
+  isArchived: boolean | undefined;
+  createdAt: Date | undefined;
+  starsCount: number | undefined;
+  forksCount: number | undefined;
+  openIssues: number | undefined;
+
+  constructor(
+    provider: ReleaseProvider,
+    identifier: string,
+    versionRegex: string | undefined = undefined,
+    iconUrl: string | undefined = undefined,
+  ) {
+    this.provider = provider;
+    this.identifier = identifier;
+    this.versionRegex = versionRegex;
+    this.iconUrl = iconUrl;
+    this.isNewRelease = false;
+    this.isStaleRelease = false;
+  }
+}
```
I would suggest to just use an object as well

Aandree5 replied:
Ok, changed with ae4ae8542a5abf2c6a3e2e848208bfb26c89af80

Aandree5 commented:
I'm currently on holiday, I should be able to look into this towards the end of the week

Meierschlumpf commented:
Sounds great, have fun on your holiday

Aandree5 commented on packages/widgets/src/_inputs/widget-multiReleaseRepositories-input.tsx:
```diff
@@ -0,0 +1,284 @@
+                    }}
+                  />
+                </Grid.Col>
+                <Grid.Col span="content">
+                  <Text c="dimmed" fw={100} size="xs">
+                    {repository.provider.name}
+                  </Text>
+                </Grid.Col>
+                <Grid.Col span="auto">
+                  <Text size="sm">{repository.identifier}</Text>
+                </Grid.Col>
```
Sorry I didn't get this, it is currently behind the identifier. Do you mean on the same grid cell wrapped by a group?

Aandree5 commented on packages/widgets/src/_inputs/widget-multiReleaseRepositories-input.tsx:
```diff
@@ -0,0 +1,284 @@
+  providers: string[];
+}
+
+const releaseEditModal = createModal<ReleaseEditProps>(({ innerProps, actions }) => {
```
Good point, fixed with 1f0268d76614dc69eaf8f2e98e6414c7885a8d61

Aandree5 commented on packages/widgets/src/_inputs/widget-multiReleaseRepositories-input.tsx:
```diff
@@ -0,0 +1,284 @@
+      setFormErrors(validation.errors);
+    },
+    [innerProps, tempRepository],
+  );
```
This is a very good point, I already had a local state, not sure why I was changing the main state `onChange` instead of only `onConfirm`. I think I had seen this issue before but got side tracked and forgot. Fixed with ed657fea1bae624581a3da111ed9cdbaafc967c5

Aandree5 commented on packages/widgets/src/_inputs/widget-multiReleaseRepositories-input.tsx:
```diff
@@ -0,0 +1,284 @@
+              "^v[0-9]+\\.[0-9]+\\.[0-9]+$",
+              "^[0-9]+\\.[0-9]+\\.[0-9]+\\.[0-9]+$",
+              "^[0-9]+\\.[0-9]+\\.[0-9]+$",
+              "^[0-9]+\\.[0-9]+$",
```
No sure what you meant here, but I just changed it to `data={Object.keys(versionRegexData)}` with 27b672cb3cec7b63bf251ace5cdc1fe15ae3d59d

Meierschlumpf replied:
That's a good solution too

Aandree5 commented on packages/request-handler/src/releases.ts:
```diff
@@ -0,0 +1,351 @@
+    if (!releasesResult.success) return undefined;
+
+    const latest: ReleaseResponse = releasesResult.data
+      .filter((result) => (input.versionRegex ? new RegExp(input.versionRegex).test(result.latestRelease) : true))
```
It was more to give freedom to people to track whichever release they prefer, but I do see your point.
Do you have a suggestion for this? 
We could either have a bigger set of available options to pick or allow a limit set of changes, maybe someting like:
- Adding a prefix
- Precision of version number, i.e. 2 for (1.2), 3 for (1.2.3), etc...
- Adding a suffix

So they could not type an actual regex value, this could potentially cover all basis (or most at least)

Meierschlumpf replied:
Yeah this prefix precision suffix version I like!

Aandree5 replied:
This is done with 4f3841f4bc29321b508d7e8cee3092c11af60827 and d11ae021da6b194632de098c58e9ac75a9f6920a

https://github.com/user-attachments/assets/da0092bd-b2f4-48e7-a0bc-73267c7e64f4

Aandree5 commented on packages/request-handler/src/releases.ts:
That is what I was trying to so with:
```
export const Providers: ProvidersProps = {
  DockerHub: new ReleaseProvider("Docker Hub", "https://cdn.jsdelivr.net/gh/selfhst/icons/svg/docker.svg"),
  Github: new ReleaseProvider("Github", "https://cdn.jsdelivr.net/gh/selfhst/icons/svg/github-dark.svg"),
  Gitlab: new ReleaseProvider("Gitlab", "https://cdn.jsdelivr.net/gh/selfhst/icons/svg/gitlab.svg"),
  Npm: new ReleaseProvider("Npm", "https://cdn.jsdelivr.net/gh/loganmarchione/homelab-svg-assets//assets/npm.svg"),
  Codeberg: new ReleaseProvider("Codeberg", "https://cdn.jsdelivr.net/gh/selfhst/icons/svg/codeberg.svg"),
};
```
Do you have a better option on where to handle this?

Aandree5 commented on packages/widgets/src/releases/component.tsx:
```diff
@@ -0,0 +1,370 @@
+
+    const repository = options.repositories.find(
+      (repository: ReleaseRepository) => repository.identifier === data.identifier,
+    );
```
Yes makes sense, fixed 9205b5a311f4e3b32133d3e1cb8bb1c7333bc7fc

Aandree5 commented on packages/widgets/src/releases/component.tsx:
```diff
@@ -0,0 +1,370 @@
+      isStaleRelease:
+        options.staleReleaseWithin !== "" ? !isDateWithin(data.latestReleaseAt, options.staleReleaseWithin) : false,
+    });
+  }
```
Couldn't find the `array.map()` in that code, but I think I got what it was, the repositores passed onto the request right?
If so, it's fixed cffada5d57394244e268dd0873af48e2833674c4

Aandree5 replied:
got it, done 6fd620f09c85e5726de58fb9f832f5808b9bce38

Aandree5 commented on packages/widgets/src/releases/component.tsx:
```diff
@@ -0,0 +1,370 @@
+              align="center"
+              onClick={() => toggleExpandedRepository(repository.identifier)}
+            >
+              <Grid.Col span={1.2}>
```
Just thought I could control the sizes better, would you prefer flex? Do you think it will improve the usability?

Aandree5 replied:
This is fixed with 36eca0b7d476be6299cf69a5e78126ffc67d37a7 and b476ffcde00d2aa62405eb563b43a616561e26f9

Aandree5 commented on packages/widgets/src/releases/component.tsx:
```diff
@@ -0,0 +1,370 @@
+          rel="noreferrer"
+        >
+          <IconExternalLink />
+          {t("openProjectPage")}
```
Done, with 62d3bbc26056cda15fbfa5e6e9ed2038c6e7e119

Aandree5 commented on packages/widgets/src/releases/component.tsx:
```diff
@@ -0,0 +1,370 @@
+              <ReactMarkdown
+                skipHtml
+                components={{
+                  a: ({ ...props }) => <span>{props.children}</span>,
```
I just thought of it as more of a security potential issue, there could be some nefarious link, I wasn't sure, also not sure if usability would improve. Was looking at that description more like a summary/overview of the changes, and if people want to follow it through they could open the link, but I can allow links if you think would be fine security wise.

Meierschlumpf replied:
I think it's okay to include links normally, what's your opinion @manuel-rw?

Aandree5 replied:
I've allowed for link to be displayed, if this changes it will be easy to fix anyway

Aandree5 commented on packages/widgets/src/releases/index.ts:
```diff
@@ -0,0 +1,67 @@
+import { optionsBuilder } from "../options";
+
+export const { definition, componentLoader } = createWidgetDefinition("releases", {
+  icon: IconCode,
```
That makes total sense, I had a different name at the beggining and forgot to update the icon. FIxed 40b559c7b7a7bb77111220da4c59b1676681cfec

Meierschlumpf commented on packages/request-handler/src/releases.ts:
```diff
@@ -0,0 +1,354 @@
+
+import { fetchWithTimeout } from "@homarr/common";
+
+import { Providers } from "../../widgets/src/releases/release-providers";
```
I think this will cause an issue on the dev server. If so I'll suggest that we just move this variable in this file and add the properties

- `detailsUrl`
- `detailsSchema`
- `releasesUrl`
- `releasesSchema`

to each of them (or at lest there where we have detail urls for example
And then you can remove your switch statement and directly use these provider definitions

Aandree5 replied:
Think this is done, made a big change with 4ce6b594f879c6174f47f3e136408099eb03f5d1

Meierschlumpf replied:
The above (from yesterday) is still pending. You'll need to move the release-providers file to this package

Aandree5 replied:
Sorry I missed this one. I will try to get onto this today. Would it be worth to move both files? The release-repository one so they are in the same place or do I just move the providers?

Aandree5 replied:
If I move `release-providers` to the `request-handler` package how do I call it from the current places? It's used in the `widgets` package which does not have the `request-handler` in the `package.json`. Would I add it or just import by folder like `../../../request-handler/src/release-providers.ts`?

Aandree5 replied:
@Meierschlumpf can you give me some help on this? Not sure what would be the best route forward

Meierschlumpf replied:
Sorry for not answering the last few days. Was not really working on Homarr

Aandree5 commented:
I've done all the required changes, Mainly the `vergionRegex` has been replaced with `versionFilter` which has the option for `prefix`, `sufix` and a version `precision` which should cover most of the requirements, this can also be adjusted easilly in the future if needed.

The provider specific code from the request handler has been moved to the respective provider object, this made more sense since the repository itself had no need to keep track of the provider information, only the `providerKey` was required to later access required information. The code does look a lot cleaner now, but let me know you thoughts on this.

There is a small issue that can arise, mainly with GitHub but I do think it could work the same for most providers anyway. They limit the numebr of requests an IP can do per hour, this can be increased by adding an authentication key to the request, my question is where would the user type this autentication code, would it be on the options like all the other widget options? Or do you think the default limit is fine for 99% of the applications? I probably only reached the limit because of development and having to refresh the data several times.

Meierschlumpf commented on packages/widgets/src/_inputs/widget-multiReleaseRepositories-input.tsx:
```diff
@@ -0,0 +1,309 @@
+            label: provider.name,
+          }))}
+          value={tempRepository.providerKey}
+          key={`${innerProps.fieldPath}.providerKey`}
```
You don't need `key` when not having an array

Meierschlumpf commented on packages/widgets/src/releases/component.tsx:
```diff
@@ -0,0 +1,366 @@
+              onClick={() => toggleExpandedRepository(repository.identifier)}
+            >
+              <MaskedOrNormalImage
+                imageUrl={repository.iconUrl ?? Providers[repository.providerKey]?.iconUrl ?? ""}
```
Make `imageUrl` able to receive undefined so we don't get errors like:

![image](https://github.com/user-attachments/assets/709b7b24-53df-40ec-9d08-46c43d5b5c3f)

Aandree5 replied:
Fixed b05537abbda7372e36ad44c516191764777dc5c8

Meierschlumpf commented on packages/widgets/src/_inputs/widget-multiReleaseRepositories-input.tsx:
```diff
@@ -0,0 +1,309 @@
+  );
+
+  const addNewItem = () => {
+    const item = {
+      provider: Providers.DockerHub,
+    };
```
Specify values for all fields. Otherwise we get errors like:

![image](https://github.com/user-attachments/assets/965c4188-85ff-47bb-87a3-69054e7f3fdc)


Because the value previously was undefined. You can obviously just use empty strings

Aandree5 replied:
Fixed d948f57f887f9b97e6ea9420eba83085fbbda69f

Aandree5 commented on packages/api/src/router/widgets/releases.ts:
```diff
@@ -0,0 +1,36 @@
+          z.object({
+            providerKey: z.string(),
+            identifier: z.string(),
+            versionRegex: z.string().optional(),
```
Think it's fixed with 0f7ecca947c4cb86a0493cb1c2c74d16e6c344b1 I still left a copy on the client side just to show a representation of how it would look, but nothing is saved. I'm not sure if we could have these 2 functions somewhere easilly accessible by both sides

Aandree5 commented on packages/request-handler/src/releases.ts:
```diff
@@ -0,0 +1,97 @@
+    const releasesResponse = await fetchWithTimeout(provider.getReleasesUrl(input.identifier));
+    const releasesResult = provider.parseReleasesResponse(await releasesResponse.json());
+
+    if (!releasesResult.success) return undefined;
```
Fixed 34900a302df859ee7b3f0f096ec1dcf50c7b1d9e

Aandree5 commented on packages/widgets/src/_inputs/widget-multiReleaseRepositories-input.tsx:
```diff
@@ -0,0 +1,309 @@
+import { useScopedI18n } from "@homarr/translation/client";
+import { MaskedOrNormalImage } from "@homarr/ui";
+
+import { IconPicker } from "../../../forms-collection/src";
```
Done 16f5621f96f469edd8dce9d12bcef0cd3397e665

Aandree5 commented:
Think all the issue have been addressed, the one I'm not too sure about is moving the regex creating to the server side, done with `formatVersionFilterRegex`. But need to have two of the same.

ajnart commented:
@Aandree5 Great work on this! Do you think you could make it so that this widget has default repositories to all of the integrations created ? 
It would speed up the setup by a lot for new users of that widget

ajnart commented on packages/widgets/src/releases/release-providers.ts:
```diff
@@ -0,0 +1,318 @@
+  Github: {
+    name: "Github",
+    iconUrl: "https://cdn.jsdelivr.net/gh/selfhst/icons/svg/github-dark.svg",
+    getDetailsUrl(identifier) {
+      const [owner, name] = identifier.split("/");
+      if (!owner || !name) {
+        return "";
+      }
+      return `https://api.github.com/repos/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`;
+    },
```
Make sure to handle some error case if the user is Chinese and Github is blocked (you can try locally by adding it to an adblocker) 

Or the API rate limiting has been hit

Meierschlumpf replied:
I agree with Andre on this, if they know that GitHub does not work, they should not use it 😉

Aandree5 replied:
The only thing, if there is any error then the repository will simply dissapear from the list (which could be difficult for the user to realise). I working changing it to leave it in the list and show the error message returned from the call, maybe on a tooltip. Should I send this on a new PR? It's not fully working yet

Aandree5 replied:
I will send on a new PR

Meierschlumpf replied:
Yeah we can show somewhere that errors occured for certain repositories, but I would say that we can put this in a separat PR

ajnart commented:
IMO the regex field must be an advanced option/hidden. I didn't immediately understand what it is useful for.
For me if current release != latest release (index 0 in the /tags endpoint) then there is a new update available

I would instead add an advanced option to "Only show update if "Major/Minor/Patch" release toggle and then you can use the regex logic internally to do this kind of compare.

Aandree5 commented:
> @Aandree5 Great work on this! Do you think you could make it so that this widget has default repositories to all of the integrations we provide ?
> 
> It would incredibly speed up the setup for new users of that widget. It's better to have too many defaults and let the user delete the ones they don't need rather than having them configure one thing again.
> 
> Also, having the option to quickly pick an icon from the supported integrations / apps the user has created would be a nice to have

@ajnart Thank you! I do think this could be a great tool to have. 
I'm not sure that having default rapositories for all integrations would be possible. I do like the idea, the more we can have automated the better. The only thing I think could work is the docker integration and trying to load the information about the containers it finds, but to be honest I'm not sure how easilly achievable this would be, from a coding prespective.

For the icons, I'm not sure, I think I see it becoming just another dropdown box with quite a few icons and having to search for a specific one anyway (just with less options).

> Make sure to handle some error case if the user is Chinese and Github is blocked (you can try locally by adding it to an adblocker)
> 
> Or the API rate limiting has been hit

I'm not sure what exaclty you mean here. I guess if a user is in China they would probably not try to add a Github provider if it is blocked. Saying that, I do agree with somehow showing something, because currently if the limit it hit, the repositories will just not load at all, maybe at the bottom it could show the number of failed/successful checks and it's message (will look into it when I get a chance).

> IMO the regex field must be an advanced option/hidden. I didn't immediately understand what it is useful for.
> For me if current release != latest release (index 0 in the /tags endpoint) then there is a new update available
> 
> I would instead add an advanced option to "Only show update if "Major/Minor/Patch" release toggle and then you can use the regex logic internally to do this kind of compare.

The `regex` field has been removed in favour of `versionFilter` which should be safer and more descriptive.
The problem here is that I don't currently know what the current version is. Specifying the `versionFilter` will just tell the system to get the latest `#.#.#` version, e.g. 1.2.3, so this will always be the latest release for that version filter. The filter just helps users narrow down on the type of version they want to follow, some repositories can show the most recent release as a `dev` or `pre-release` release, and the user is only interested in a production release, or they are after the `-alpine` release for example. Also, the latest release will often just be `latest` which doesn't help much, other times I've seen it being just `3` when the second would be `3.2` and the third `3.2.1` and actually the third would be more acurate.
With the last one, not knowing the current release makes this really difficult, plus all of the most recent releases will be at the top and highlighted this should tell the user quickly if/any repository has had a release recently.

ajnart commented:
I didn't take into account the fact that the current version is unknown. This makes a lot of sense now

Aandree5 commented:
I see what you mean, think it could be a good idea to have such defaults

Aandree5 commented:
All issues should now be fixed

Meierschlumpf commented:
So the issue with the `Input is too big for a single dispatch` is also working again?

Aandree5 commented:
Yes it appears so, I tested with a new database and it all works fine with 3 widgets with 9 repositories each. Not sure if the issue was some of the data that had been previously saved, when I changed some fields around

Meierschlumpf commented:
@ajnart do you agree that adding the defaults can be done in a second pr and we can merge this pr before fridays release?

ajnart commented:
> @ajnart do you agree that adding the defaults can be done in a second pr and we can merge this pr before fridays release?

100%. We can add it to the task list of adding defaults to pretty much everything

Meierschlumpf commented:
@Aandree5 thank you very much for this contribution ♥️, I really like how the widget works now!
Let's wait for another @homarr-labs/maintainers for a second approval

manuel-rw commented:
Reviewing... 🕔

manuel-rw commented:
![releases-widget](https://github.com/user-attachments/assets/15696fec-1896-4948-aca6-b0ad40b84e5a)

I tried adding our GitHub repo. That worked fine. But after that, I tried to add more and the widget refused to accept any button clicks:
![releases-widget](https://github.com/user-attachments/assets/76c512e5-8c46-4927-93c4-ab5bf8ca8e92)

Either some error message should be displayed (e.g. validation) or it should save and close.
Can you have a look before I review code?

Aandree5 commented:
Yes good point, fixed with a86005e554d74c738c928da99b584286ded8d69e and e5c0921debe5e111593ea1ec3c11fb33c667344a, made is so the modal auto-opens when the user clicks "Add.." which should help with usability, and added a message to the repository item on the list to show if any errors are present

https://github.com/user-attachments/assets/abdaa7c8-a28b-4ece-aba2-7d492cf78910

Meierschlumpf commented on packages/widgets/src/_inputs/widget-multiReleasesRepositories-input.tsx:
```diff
@@ -78,7 +98,7 @@ export const WidgetMultiReleasesRepositoriesInput = ({
     });
   };
-
+  console.log(form.errors);
```
Remove console.log

Aandree5 replied:
Removed 3a0e7aadd645f780d017ab1e4704a646f5bc6774

manuel-rw commented:
Due to absences, I will merge this pull request on behalf of @Meierschlumpf .

manuel-rw commented:
Wow this is one heck of a commit :D Thanks for your contribution, we **really** appreciate this!
